File: 49GroupAnagrams.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> umap;
        string temp;
        vector<vector<string>> ordered;
        for (auto x: strs){
            temp = x;
            sort(temp.begin(),temp.end());
            umap[temp].push_back(x);
        }
        for(auto mappair: umap){
            ordered.push_back(mappair.second);
        }
        return ordered;
    }
};

int main(){
    Solution tester;
    vector<string> strs = {"eat","tea","tan","ate","nat","bat"};
    vector<vector<string>> result = tester.groupAnagrams(strs);
    for ( auto& row: result){
        for(auto& entry: row){
            cout << entry << ",";
        }
        cout << endl;
    }
}

File: 49GroupAnagrams.py
class Solution(object):
    def groupAnagrams(self, strs):
        """
        :type strs: List[str]
        :rtype: List[List[str]]
        """
        dict = {}
        for currstr in strs:
            sorted_str = ''.join(sorted(currstr))
            if sorted_str in dict:
                dict[sorted_str].append(currstr)
            else:
                dict[sorted_str] = [currstr]
        return list(dict.values())

if __name__ == "__main__":
    tester = Solution()
    test_cases = [
        (["eat","tea","tan","ate","nat","bat"]),
        (["a"]),
        ([])
    ]
    for case in test_cases:
        result = tester.groupAnagrams(case)
        print(f"groupAnagrams({case}) = {result}")